### ISSUE TITLE
Tensor2Bin abandons the tensor stream on the first variant-free batch (sparse truth)

### ISSUE THREAD

**Issue by onsiteseq** [HKU-BAL/Clair3#468]
#Tensor2Bin abandons the tensor stream on the first variant-free batch (sparse truth)

**Suggested title:** `Tensor2Bin/get_training_array: empty shuffle batch after maximum_non_variant_ratio filtering is treated as end-of-input — truncated bins + BrokenPipe in CreateTensor* (sparse truth sets)`

**Labels:** bug, training

---

## Summary

`Tensor2Bin` (via `clair3/utils.py::get_training_array`) consumes the tensor
stream in batches of `shuffle_bin_size = 50000`. With
`--maximum_non_variant_ratio` set, `_filter_non_variants` drops **all**
non-variant tensors from any batch that contains **zero truth variants**,
yielding an empty batch — and the read loop treats an empty batch as
end-of-input:

```python
# clair3/utils.py, get_training_array (v1.0.10, line 456; main: line 537)
            if X is None or not len(X):
                break
```

Tensor2Bin then exits normally (`Finish!`), the producer
(`CreateTensorFullAlignment` / `CreateTensorPileup`) gets `BrokenPipeError`
on its next write, and the output bin is **silently truncated** — all
remaining genomic regions of the contig (including any truth variants in
later batches) never make it into the training data.

## Why it stays invisible on human data — and fires on ours

Batches are formed in **genomic order**. With a dense truth set (human
training, hundreds of thousands of variants), every 50k-candidate batch
contains variants. We fine-tune on a nearly isogenic fungal sample
(*Pyricularia oryzae* Guy11 vs the 70-15 reference, ~3.4k truth variants
genome-wide): stretches of 50k consecutive AF-based candidates without a
single truth variant are the norm, so the first variant-free batch terminates
the input mid-contig.

Tell-tale numbers in our logs: `Compressed 42/449999`, `954/249999`,
`0/49999` — i.e. exactly `N × 50000 − 1` tensors consumed before the stop;
contigs with variants in every batch read to completion.

## Affected versions

Confirmed in tag **v1.0.10**; the same `break` is still in `main`
(`clair3/utils.py`, line 537).

## Proposed fix

An empty post-filter batch is not end-of-input — skip it and keep reading:

```diff
             try:
                 X, total, completed = next(bin_g)
             except StopIteration:
                 completed = True

-            if X is None or not len(X):
-                break
+            if X is None:
+                break
+            if not len(X):
+                # a batch with zero truth variants is emptied by
+                # _filter_non_variants(); keep reading instead of
+                # abandoning the rest of the tensor stream
+                continue
```

With this change every truth variant with read support is kept (variants are
never dropped by `_filter_non_variants`), empty batches simply contribute
nothing, and the producer no longer dies with BrokenPipe.

## Edge case worth a look

If a whole contig has zero usable truth variants, the resulting bin contains
0 tensors (legitimate under a variant:non-variant ratio). Callers
(`MergeBin`, `Train`) should tolerate or skip such empty bins.

Happy to open a PR with the fix if that helps.
